src/main/scala/org/constellation/gossip/GossipService.scala

```diff
@@ -1,47 +1,127 @@
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import org.constellation.ConfigUtil
+import org.constellation.gossip.bisect.bisectA
+import org.constellation.gossip.sampling.{GossipPath, PeerSampling}
+import org.constellation.gossip.state.{GossipMessage, GossipMessagePathTracker}
 import org.constellation.infrastructure.p2p.PeerResponse.PeerClientMetadata
 import org.constellation.p2p.Cluster
-import org.constellation.schema.{Id, NodeState}
+import org.constellation.schema.Id
 
-abstract class GossipService[F[_], A](selfId: Id, peerSampling: PeerSampling[F, Id], cluster: Cluster[F])(
-  implicit val F: Concurrent[F]
+import scala.concurrent.duration._
+
+abstract class GossipService[F[_]: Parallel, A](
+  selfId: Id,
+  peerSampling: PeerSampling[F],
+  cluster: Cluster[F],
+  messageTracker: GossipMessagePathTracker[F, A]
+)(
+  implicit F: Concurrent[F],
+  T: Timer[F]
 ) {
+  protected val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  private val spreadRequestTimeout = ConfigUtil.getDurationFromConfig("gossip.spread-request-timeout", 10 seconds)
+
+  /**
+    * How to spread the message to peer
+    */
+  protected def spreadFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Unit]
+
+  /**
+    * How to check if message was received by peer
+    */
+  protected def validationFn(peerClientMetadata: PeerClientMetadata, message: GossipMessage[A]): F[Boolean]
+
+  // TODO: Check if bisect is efficient enough for error handling Gossip
+  def recover(message: GossipMessage[A]): F[Unit] = {
+    val sequence = message.path.toIndexedSeq
 
-  protected def spread(data: A, fanout: Int)(spreadFn: (PeerClientMetadata, GossipMessage[A]) => F[Unit]): F[Unit] =
     for {
-      messages <- peerSampling
-        .selectPaths(fanout)
-        .map(_.map(GossipPath(_)).map(GossipMessage(data, _)))
+      _ <- messageTracker.fail(message.path.id)
+      failureNode <- bisectA(
+        (id: Id) => getClientMetadata(id).flatMap(validationFn(_, message))
+      )(sequence)
+      sequenceToRetry = sequence.dropWhile(failureNode.contains)
+      clients <- sequenceToRetry.toList.traverse(getClientMetadata)
+      _ <- clients.traverse(spreadFn(_, message)) // TODO: Force trigger peer healthcheck if it fails
+      _ <- messageTracker.remove(message.path.id)
+    } yield ()
+  }
 
-      _ <- messages.toList.traverse(spreadToNext(_)(spreadFn))
+  /**
+    * Start spreading the message to fanout
+    */
+  def spread(data: A): F[Unit] =
+    for {
+      _ <- logger.debug(s"Starting spreading of data: $data")
+      messages <- peerSampling.selectPaths
+        .map(_.map(GossipMessage(data, _)))
+      _ <- messages.map { msg =>
+        for {
+          _ <- messageTracker.start(msg)
+          _ <- logger.debug(s"Started path ${msg.path.id}: ${msg.path.toIndexedSeq.map(_.short).mkString(" -> ")}")
+          _ <- (for {
+            _ <- spreadToNext(msg)
+            _ <- T.sleep(getPathTimeout(msg.path))
```

Comment:
Which thread pool it executes on? On unbounded? If no, please eval the sleep on unbounded explicitly